server/src/main/java/org/elasticsearch/indices/IndexingMemoryController.java

```diff
@@ -393,14 +429,15 @@ private void runUnlocked() {
                         largest.shard.shardId(),
                         ByteSizeValue.ofBytes(largest.bytesUsed)
                     );
-                    writeIndexingBufferAsync(largest.shard);
+                    enqueueWriteIndexingBuffer(largest.shard);
                     totalBytesUsed -= largest.bytesUsed;
```

Comment:
What I've observed in practice is that we check the total size of the indexing buffer so frequently (every indexing_buffer_size/128 bytes written to the translog across shards, so indexing buffer typically increased by  less than 1% since the previous check) that this method only ever collects a single shard to flush at once. So we wouldn't hit the issue that you are describing?